### ISSUE TITLE
Discrepancy in Type Narrowing with Union Types and TypeVars in Pyright

### ISSUE THREAD

**Issue by ConsciousCode** [microsoft/pyright#7111]
#### Issue Description:
I've encountered an issue in Pyright where there is a discrepancy in type narrowing between two similar function signatures using TypeVars and union types. Notably, Pyright issues a warning for a function signature (`narrow_success_2`) that behaves differently from its suggested "fix" (`narrow_success_1`). This discrepancy is crucial in cases where precise type narrowing from union types is needed.

#### Environment:
- Pyright version: 1.1.347 (VSCodium extension)
- Python version: 3.12
- Operating System: Linux x86_64 (Ubuntu 22.04.3 LTS - jammy)

#### Code to Reproduce:
```python
class Response: ...
class ErrorResponse(Response): ...
class LoginError(ErrorResponse): ...
class LoginResponse(Response): ...

def narrow_success_1[T: Response](x: T|ErrorResponse) -> T: ...

# Pyright warning: "TypeVar "E" appears only once in generic function signature. Use "ErrorResponse" instead"
def narrow_success_2[T: Response, E: ErrorResponse](x: T|E) -> T: ...

# Sample implementation of narrow_success
def narrow_success_impl(x):
    if isinstance(x, ErrorResponse):
        raise RuntimeError()
    return x

def login() -> LoginResponse|LoginError: ...

reveal_type(narrow_success_1(login())) # LoginResponse|LoginError
reveal_type(narrow_success_2(login())) # LoginResponse
```

#### Behavior:
`narrow_success_1` retaining the union type `LoginResponse|LoginError` is unexpected, but can be justified as `T` binding to `LoginResponse|LoginError` (since both are subclasses of `Response`), with the `ErrorResponse` part being ignored. `narrow_success_2`, which Pyright suggests should be simplified to `narrow_success_1`, actually narrows the type to `LoginResponse`, demonstrating different type hinting behaviors.

#### Concern:
Either the inferred type of `narrow_success_1` is out of spec and should narrow the union type, or the warning for `narrow_success_2` should be reconsidered, as it has a legitimate use case that cannot be achieved with the suggested simplification. The distinct behavior of `narrow_success_2` is crucial for precise type narrowing in certain scenarios, which is not possible with the `narrow_success_1` signature.

### Context:
I found this discrepancy while writing code using the `matrix-nio` library. Ordinarily I would have a `SuccessResponse` subclass so the two branches are clearly distinguished, as I've run into issues like this before, but since it's a third party library I can't do that. I wanted to reduce boilerplate by writing a `unwrap` function to narrow the type of the returns:
```python
response = unwrap(await self.client.room_create(...))
return response.room_id # Pyright error: Cannot access member "room_id" for type "RoomCreateError"
```

This works with `narrow_success_2`, and I'd much prefer the one warning on `E` than the error on every access to every unwrapped response.

### Other options:
```python
@overload
def narrow_success_3(x: ErrorResponse) -> NoReturn: ...
@overload
def narrow_success_3[T: Response](x: T) -> T: ...

def narrow_success_3(x):
    if isinstance(x, ErrorResponse):
        raise RuntimeError()
    return x

reveal_type(narrow_success_3(login())) # LoginResponse|LoginError
```

```python
@overload
def narrow_success_4[E: ErrorResponse](x: E) -> NoReturn: ...
@overload
def narrow_success_4[T: Response](x: T) -> T: ...

def narrow_success_4(x):
    if isinstance(x, ErrorResponse):
        raise RuntimeError()
    return x

reveal_type(narrow_success_4(login())) # LoginResponse|LoginError
```

**Comment by erictraut** [comment]
I'll start by making a small point about terminology. You used the phrase "type narrowing" in the title, but the issue has nothing to do with type narrowing. It's related to TypeVar constraint solving.

Constraint solving is the process by which a type checker "solves" type variables for a function call. The goal is to find some type for each TypeVar that satisfies all of the constraints. Constraints can come from upper bounds, variance, and the types of arguments provided in the call expression.

There are often multiple valid ways for a constraint solver to "solve" type variables. No single solution is necessarily more "correct" than another long as all constraints are met. Pyright's constraint solver includes a bunch of heuristics that attempt to guide it toward a solution that is most useful in most circumstances. As with any heuristics, there are situations where they will fail or will produce an answer that is suboptimal for your particular use case.

When you include a "naked" type variable in a union (e.g. in the type annotation `T | ErrorResponse` or `T | E`), you create more freedom for the constraint solver, and that increases the chances that you'll see different behaviors. This explains the difference that you're seeing between `narrow_success_1` and `narrow_success_2`. Neither of these are "out of spec", and pyright is working as designed in both cases.

The warning that you see for `narrow_success_2` was added to pyright because many people do not understand how to use TypeVars and tend to use them incorrectly. If you don't like the warning or think that it's not helpful in this case, you can suppress it or disable the rule entirely for your project.

In summary, pyright is working correctly here, so I don't consider this a bug.
